packages/ics721/src/ibc_packet_receive.rs

```diff
@@ -94,57 +121,117 @@ pub(crate) fn receive_ibc_packet(
+        )?;
+    let is_redemption = if !redemption_or_create.0.is_empty() && !redemption_or_create.1.is_empty()
+    {
+        // All token ids in the transfer must be either a redeption or creation
+        // they can't be both, if they are both something is wrong.
         return Err(ContractError::InvalidTransferBothActions);
-    }
+    } else if !redemption_or_create.0.is_empty() {
+        true
+    } else if !redemption_or_create.1.is_empty() {
+        false
+    } else {
+        // This should never happen, as we must have at least 1 of the above actions
+        return Err(ContractError::InvalidTransferNoAction);
+    };
+
+    let receiver = deps.api.addr_validate(&data.receiver)?;
+    let voucher_and_channel_messages = match is_redemption {
+        true => {
+            let redemption = VoucherRedemption {
+                class: Class {
+                    id: local_class_id.clone(),
+                    uri: data.class_uri.clone(),
+                    data: data.class_data.clone(),
+                },
+                token_ids: redemption_or_create.0,
+            };
+            let redeem_outgoing_channels: Vec<(ClassId, TokenId)> = redemption
+                .token_ids
+                .clone()
+                .into_iter()
+                .map(|token_id| (local_class_id.clone(), token_id))
+                .collect();
+            let redeem_outgoing_channels_msg = WasmMsg::Execute {
+                contract_addr: env.contract.address.to_string(),
+                msg: to_json_binary(&ExecuteMsg::Callback(
+                    CallbackMsg::RedeemOutgoingChannelEntries(redeem_outgoing_channels),
+                ))?,
+                funds: vec![],
+            };
+            (
+                redemption.into_wasm_msg(env.contract.address.clone(), receiver.to_string())?,
+                redeem_outgoing_channels_msg,
+            )
+        }
+        false => {
+            let creation = VoucherCreation {
+                class: Class {
+                    id: local_class_id.clone(),
+                    uri: data.class_uri.clone(),
+                    data: data.class_data.clone(),
+                },
+                tokens: redemption_or_create.1,
+            };
+            let add_incoming_channels: Vec<((ClassId, TokenId), String)> = creation
+                .tokens
+                .clone()
+                .into_iter()
+                .map(|token| {
+                    (
+                        (local_class_id.clone(), token.id),
+                        packet.dest.channel_id.clone(),
+                    )
+                })
+                .collect();
+            let add_incoming_channels_msg = WasmMsg::Execute {
+                contract_addr: env.contract.address.to_string(),
+                msg: to_json_binary(&ExecuteMsg::Callback(
+                    CallbackMsg::AddIncomingChannelEntries(add_incoming_channels),
+                ))?,
+                funds: vec![],
+            };
```

Comment:
... CallbackMsg::AddIncomingChannelEntries which then will be added to sub message.